libpod/sqlite_state.go

```diff
@@ -0,0 +1,2535 @@
+		}
+		defer rows.Close()
+
+		for rows.Next() {
+			var rawJSON string
+			if err := rows.Scan(&rawJSON); err != nil {
+				return nil, fmt.Errorf("scanning container from database: %w", err)
+			}
+
+			ctr := new(Container)
+			ctr.config = new(ContainerConfig)
+			ctr.state = new(ContainerState)
+			ctr.runtime = s.runtime
+
+			if err := json.Unmarshal([]byte(rawJSON), ctr.config); err != nil {
+				return nil, fmt.Errorf("unmarshalling container config: %w", err)
+			}
+
+			ctrs = append(ctrs, ctr)
+		}
+	}
+
+	for _, ctr := range ctrs {
+		if err := finalizeCtrSqlite(ctr); err != nil {
+			return nil, err
+		}
+	}
+
+	return ctrs, nil
+}
+
+// GetNetworks returns the networks this container is a part of.
+func (s *SQLiteState) GetNetworks(ctr *Container) (map[string]types.PerNetworkOptions, error) {
+	if !s.valid {
+		return nil, define.ErrDBClosed
+	}
+
+	if !ctr.valid {
+		return nil, define.ErrCtrRemoved
+	}
+
+	// if the network mode is not bridge return no networks
+	if !ctr.config.NetMode.IsBridge() {
+		return nil, nil
+	}
+
+	cfg, err := s.getCtrConfig(ctr.ID())
+	if err != nil {
+		if errors.Is(err, define.ErrNoSuchCtr) {
+			ctr.valid = false
+		}
+		return nil, err
+	}
+
+	return cfg.Networks, nil
+}
+
+// NetworkConnect adds the given container to the given network. If aliases are
+// specified, those will be added to the given network.
+func (s *SQLiteState) NetworkConnect(ctr *Container, network string, opts types.PerNetworkOptions) error {
+	return s.networkModify(ctr, network, opts, true, false)
+}
+
+// NetworkModify will allow you to set new options on an existing connected network
+func (s *SQLiteState) NetworkModify(ctr *Container, network string, opts types.PerNetworkOptions) error {
+	return s.networkModify(ctr, network, opts, false, false)
+}
+
+// NetworkDisconnect disconnects the container from the given network, also
+// removing any aliases in the network.
+func (s *SQLiteState) NetworkDisconnect(ctr *Container, network string) error {
+	return s.networkModify(ctr, network, types.PerNetworkOptions{}, false, true)
+}
+
+// GetContainerConfig returns a container config from the database by full ID
+func (s *SQLiteState) GetContainerConfig(id string) (*ContainerConfig, error) {
+	if len(id) == 0 {
+		return nil, define.ErrEmptyID
+	}
+
+	if !s.valid {
+		return nil, define.ErrDBClosed
+	}
+
+	return s.getCtrConfig(id)
+}
+
+// AddContainerExitCode adds the exit code for the specified container to the database.
+func (s *SQLiteState) AddContainerExitCode(id string, exitCode int32) (defErr error) {
+	if len(id) == 0 {
+		return define.ErrEmptyID
+	}
+
+	if !s.valid {
+		return define.ErrDBClosed
+	}
+
+	tx, err := s.conn.Begin()
+	if err != nil {
+		return fmt.Errorf("beginning transaction to add exit code: %w", err)
+	}
+	defer func() {
+		if defErr != nil {
+			if err := tx.Rollback(); err != nil {
+				logrus.Errorf("Rolling back transaction to add exit code: %v", err)
+			}
+		}
+	}()
```

Comment:
The begin/rollback/exec/commit boilerplate really strongly suggests to me the need for an `execInTransaction(command string, params any...) error` helper that implements that once instead of copy&pasting.

A lower-level `execFnInTransaction(func () error) error` that just encapsulates the begin/rollback/commit parts, allowing the caller to perform multiple operations, might also pay off.